From db0fc617585073065afb1aff3c4e6f5c42748548 Mon Sep 17 00:00:00 2001
From: Colin Watson <cjwatson@debian.org>
Date: Thu, 28 Mar 2024 13:21:47 +0000
Subject: [PATCH] Fix reference-counting in set_event_filter
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This previously resulted in the following test failure with Perl 5.38.2
on Debian unstable (https://bugs.debian.org/1064761):

    t/core_events.t ................. 1/? Can't use an undefined value as a subroutine reference during global destruction.
    t/core_events.t ................. Dubious, test returned 22 (wstat 5632, 0x1600)
    All 691 subtests passed
            (less 1 skipped subtest: 690 okay)

Signed-off-by: Petr Písař <ppisar@redhat.com>
---
 src/Core/Events.xs | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/Core/Events.xs b/src/Core/Events.xs
index faa7cfe6..2da2adef 100644
--- a/src/Core/Events.xs
+++ b/src/Core/Events.xs
@@ -102,7 +102,8 @@ void
 events_set_event_filter(callback)
 	SV* callback
 	CODE:
-		eventfiltersv = callback;
+		SvREFCNT_dec(eventfiltersv);
+		eventfiltersv = SvREFCNT_inc(callback);
 		SDL_SetEventFilter((SDL_EventFilter) eventfilter_cb);
 
 AV *
-- 
2.44.0

